packages/@react-spectrum/menu/src/SafeTriangle.tsx

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import React, {MutableRefObject, ReactElement, useEffect, useState} from 'react';
+
+interface SafeTriangleProps {
+  /** Ref for the submenu. */
+  subMenuRef: MutableRefObject<any>,
+  /** Ref for the submenu's trigger element. */
+  triggerRef: MutableRefObject<HTMLLIElement>
+}
+
+/**
+ * Allows the user to move their pointer to the sub-menu without it closing
+ * due to the mouse leaving the trigger element.
+ */
+export function SafeTriangle(props: SafeTriangleProps): ReactElement {
+  let {
+    subMenuRef,
+    triggerRef
+  } = props;
+  let [mousePosition, setMousePosition] = useState<{mouseX: number, mouseY: number}>({mouseX: null, mouseY: null});
+  let {mouseX, mouseY} = mousePosition;
+  let updateMousePosition = (e: MouseEvent) => {
+    setMousePosition({mouseX: e.clientX, mouseY: e.clientY});
+  };
+
+  useEffect(() => {
+    window.addEventListener('mousemove', updateMousePosition);
+
+    return () => window.removeEventListener('mousemove', updateMousePosition);
+  }, []);
+
+  let x = 0;
+  let y = 0;
+  let width = 0;
+  let height = 0;
+  let direction: 'left' | 'right';
+  let subMenuNode = subMenuRef.current?.UNSAFE_getDOMNode() as HTMLElement;
+  if (subMenuNode) {
+    let rect = subMenuNode.getBoundingClientRect();
```

Comment:
I see a lot of reading the dom in this render, this should all be moved inside an effect and the result should be stored in state

I think doing that may help with the flicker that happens some times where the triangle is the wrong size